src/MPAGSCipher/PlayfairCipher.cpp

```diff
@@ -0,0 +1,212 @@
+    key_ = key;
+
+    // Append the alphabet
+
+    key_ += alphabet_;
+
+    // Make sure the key is upper case
+
+    std::string upperCaseKey;
+    std::transform(key_.begin(), key_.end(), std::back_inserter(upperCaseKey), ::toupper);
+
+    // Remove non-alpha characters
+
+    auto iter_alpha = std::remove_if(upperCaseKey.begin(), upperCaseKey.end(), 
+                               [] (const char testChar) {return !isalpha(testChar);} );
+    upperCaseKey.erase(iter_alpha, upperCaseKey.end());
+
+    // Change J -> I
+
+    auto findJ = [] (char testChar) {
+        if (testChar == 'J')
+            return 'I';
+        else
+            return testChar;
+    };
+    std::string removedJKey;
+    std::transform(upperCaseKey.begin(), upperCaseKey.end(), std::back_inserter(removedJKey), findJ);
+
+    // Remove duplicated letters
+
+    std::string encounteredChars;
+    auto checkDuplicate = [&] (char testChar) {
+        if (encounteredChars.find(testChar) < encounteredChars.length()) {
+            // testChar is in encounteredChars
+            return true;
+        }
+        else {
+            // testChar is not in encounteredChars
+            encounteredChars.push_back(testChar);
+            return false;
+        }
+    };
+    auto iter_duplicate = std::remove_if(removedJKey.begin(), removedJKey.end(), 
+                                         checkDuplicate);
+    removedJKey.erase(iter_duplicate, removedJKey.end());
+    
+    // Store the coords of each letter
+    // Store the playfair cipher key map
+    for (size_t i = 0; i < removedJKey.size(); i++) {
+        std::pair<size_t, size_t> coords{floor(i / 5), i % 5};
+        letterToCoordsMap[removedJKey[i]] = coords;
+        coordsToLetterMap[coords] = removedJKey[i];
+    }
+}
+
+std::string PlayfairCipher::applyCipher(const std::string& inputText,
+                                        const CipherMode cipherMode ) const {
+
+    if (key_ == "" + alphabet_) {
+        return inputText;
+    }
+    
+    // Change J → I
+    auto findJ = [] (char testChar) {
+        if (testChar == 'J')
+            return 'I';
+        else
+            return testChar;
+    };
+    std::string removedJInput;
+    std::transform(inputText.begin(), inputText.end(), std::back_inserter(removedJInput), findJ);
+
+    // If repeated chars in a digraph add an X or Q if XX
+
+    std::string preparedInput = "";
+    size_t inputSize = removedJInput.size();
+
+    for (size_t i = 0; i < inputSize; i += 2) {
+
+        char char1 = removedJInput[i];
+        if (i + 1 == inputSize) {
+            preparedInput += char1;
+            break;
+        }
+        char char2 = removedJInput[i + 1];
+
+        if (char1 == char2) {
+            preparedInput += char1; 
+            if (char1 == 'X') {
+                preparedInput += 'Q';
+            }
+            else {
+                preparedInput += "X";
+            }
+            preparedInput += char1; 
+        }
```

Comment:
You don't want line 109, because by doing that the string now has an character all on its own that isn't part of a digraph.
Instead, you want to decrement the loop index so that the next iteration of the loop starts with what is currently `char2` as its `char1`:
```suggestion
        if (char1 == char2) {
            preparedInput += char1; 
            if (char1 == 'X') {
                preparedInput += 'Q';
            }
            else {
                preparedInput += "X";
            }
            --i; 
        }
```